### ISSUE TITLE
[NextJS] Recoil does not work as expected in latest nightly - useSyncExternalStore

### ISSUE THREAD

**Issue by gurkerl83** [facebookexperimental/Recoil#1526]
The following error appears when trying the latest nightly build with the React 18 RC-0 in concurrent mode. In Recoils latest version, the error does not occur.

The git commit of the build of Recoil nightly build used https://github.com/facebookexperimental/Recoil/commit/81eb25797004f477e892841e5ffef7c752a47e93

**Error & Stack-Trace**

**Error:** 
Missing getServerSnapshot, which is required for server-rendered content. Will revert to client rendering.
This error happened while generating the page. Any console logs will be displayed in the terminal window.

**Call Stack**
Object.useSyncExternalStore
.../node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js (5055:11)
useSyncExternalStore
.../node_modules/react/cjs/react.development.js (1701:21)
Object.useRecoilValueLoadable_SYNC_EXTERNAL_STORE [as SYNC_EXTERNAL_STORE]
.../node_modules/recoil/cjs/recoil.js (4919:10)
useRecoilValueLoadable
.../node_modules/recoil/cjs/recoil.js (5124:24)
useRecoilValue
.../node_modules/recoil/cjs/recoil.js (5140:20)
TocLink
webpack-internal:///./packages/page/layout/src/components/toc/TocLink.tsx (45:84)
renderWithHooks
.../node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js (5445:16)
renderIndeterminateComponent
.../node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js (5518:15)
renderElement
.../node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js (5734:7)
renderNodeDestructive
.../node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js (5873:11)

**Update:** The app where I see the error is based on NextJs (latest version). Taking a quick search in the NextJs repo, I found they rely on the library react-server-dom-webpack in version 0.0.0-experimental-13455d26d-20211104, which is about two months old. The library gets used in a so-called "compiled" directory; it seems NextJs does not support arbitrary versions and features after all. The following line mentions the hook useSyncExternalStore is unsupported. Does that mean the problem is not Recoil but NextJs-related? 

https://github.com/vercel/next.js/blob/1ee13baf000ab391e6017c68eb4770768a8e7674/packages/next/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-writer.browser.development.server.js#L873 

Thx!

**Comment by drarmstr** [comment]
We've been testing with React 18 RC0, but not with NextJS. Do you have a reproducer test case?

As a fallback, you can disable the recoil_sync_external_store feature in Recoil_gks.js.

**Comment by gurkerl83** [comment]
I have created a reproducer; here it is https://github.com/gurkerl83/Recoil-Nighlty-NextJs
It is a tiny NextJs app with only one hook, "useRecoilValue," involved. 

Note: I had some problems with the nightly installation because a recently added post-install script failed upon installation. It has to be removed before releasing 0.6; otherwise, consumers of recoil are running into an error that post-install can not finish.

Check out the code, then run the following commands
1. yarn => yarn berry gets used, already packaged in the app
2. yarn dev => to start the development server on localhost:3000
3. yarn start => optionally, which tries to create a production build but fails due to the error described in this issue.

I hope this helps to reproduce; Thx!

**Comment by drarmstr** [comment]
First trying to reproduce your test case working with Recoil 0.5.0 so I changed the package.json to reference that version instead of nightly.  `yarn install` works but `yarn dev` fails with an `ERR+PACKAGE_PATH_NOT_EXPORTED` error.  I'm not familiar with NextJS if this is related to that?

Trying to reproduce the failure with the nightly build I see the error you are referencing about the missing gen-flow-files directory related to a postinstall step.  (a very annoying hack, btw)  Though, perhaps that error isn't actually an issue because the nightly branch is already built, so actually building it again shouldn't be necessary.  Either way, I don't think this would be blocking 0.6 release as it's related to the nightly branch, not using an NPM package.

But all that aside, your hypothesis may be correct that NextJS simply doesn't support useSyncExternalStore().  Can you validate it disabling the `recoil_sync_external_store` feature in `Recoil_gkx.js` resolves the issue?  If so, we could disable that feature when appropriate, but the question is exactly what to test for which environment is not supported.  Is this NextJS specific or more broadly SSR?

**Comment by gurkerl83** [comment]
I disabled recoil_sync_external_store in the es-based output file, and **the error disappears**, so this option seems to work. 

A nice thing would be a documentation of what parts of React subsystems, Recoil is actually relying upon in 0.6 and before.

But what does that mean for the NextJs support or other projects doing server-side rendering (SSR) or static site generation (SSG), which is another option NextJS supports and other vendors are doing similar things, e.g., Gatsby. 

As I mentioned earlier, NextJs does have pre-compiled dependencies, and they use a version of react-server-dom-webpack, which is about two months old. I follow the React development a bit, and I believe to remember there was something about this timeframe where this hook was discussed or even introduced. 

I opened an issue in the Next repo about this problem, asking if it is possible to bump the version of this dependency to somewhere a version that fits the React RC-0 release.

**Comment by drarmstr** [comment]
Thanks for confirming that!  

If bumping the NextJS dependencies doesn't resolve this then yeah, we'll want to disable using `useSyncExternalStore()` automatically so users won't have to think about this.  My question then is just if this is specific to NextJS or if we should revert for all SSR.  I'll confirm this with the React team also.  Ideally let's resolve this before releasing 0.6.

Please link any related issues or discussions in the NextJS repository here.

**Comment by gurkerl83** [comment]
Looking at the head of the stack trace from above and looking into Reacts sources, I found a wired comment with a todo. The other functions named hooks execute something like that resolveCurrentlyRenderingComponent..., but useSyncExternalStore does not?

https://github.com/facebook/react/blob/fe419346da0de40858d5a5c5992c29a1e0b79bb5/packages/react-dom/src/server/ReactPartialRendererHooks.js#L473

```
// TODO Decide on how to implement this hook for server rendering.
// If a mutation occurs during render, consider triggering a Suspense boundary
// and falling back to client rendering.
function useMutableSource<Source, Snapshot>(
  source: MutableSource<Source>,
  getSnapshot: MutableSourceGetSnapshotFn<Source, Snapshot>,
  subscribe: MutableSourceSubscribeFn<Source, Snapshot>,
): Snapshot {
  resolveCurrentlyRenderingComponent();
  return getSnapshot(source._source);
}

function useSyncExternalStore<T>(
  subscribe: (() => void) => () => void,
  getSnapshot: () => T,
  getServerSnapshot?: () => T,
): T {
  if (getServerSnapshot === undefined) {
    throw new Error(
      'Missing getServerSnapshot, which is required for ' +
        'server-rendered content. Will revert to client rendering.',
    );
  }
  return getServerSnapshot();
}
```

This is wild, maybe somehow it is React?

**Comment by drarmstr** [comment]
Ahh, looks like we might need to provide a server snapshot; let me inquire into that.

**Comment by gurkerl83** [comment]
Seems the argument for getServerSnapshot is missing in the call here,

https://github.com/facebookexperimental/Recoil/blob/7855d02f974e7fce4defee23ecd0ff4a083a0511/packages/recoil/hooks/Recoil_Hooks.js#L350

**Comment by drarmstr** [comment]
@gurkerl83  - If you'd like to test with #1530

**Comment by gurkerl83** [comment]
I can test it ones it is merged and a nightly with the changes is available, otherwise I have to build and link it with the reproducer.

**Comment by gurkerl83** [comment]
@drarmstr **I have tested the branch, and the error is gone!** Both the dev and production build of the reproducer were **successful**. 

I checked out the merge requests branch, built it, and linked it with the reproducer. Of course, the typical things were to expect, such as different react versions of the linked library and the one used in the app. I decided to set a baseline for the React version used in the experiment to 18.0.0-RC.0, and the results described appeared a successful build; I believe the other recoil hooks should work as expected.

Thx!
